### ISSUE TITLE
IntersectRay does not account for sub-frame movement

### ISSUE THREAD

**Issue by TheFacto** [godotengine/godot#17609]
**Godot version:**

* v3.0.2
* Using C#

**OS/device including version:**
* Windows 10
* Intel i7

**Issue description:**
When using `IntersectRay()`, it appears that it is using the positions of collidable objects at the start of the `_PhysicsProcess()`. I have a case where I first move an object via `SetPosition()` and subsequently call `IntersectRay()` and expect that to collide with the object at its new position. I realize this may not be worded clearly, so for sake of clarity, allow me to bullet-ize these statements.
* I have a collidable object, let's say `Platform` at position `(0, 0)`
* I move `Platform` to position `(0, 10)`
* I have a player that calls `IntersectRay()` to check where a `Platform` is
* The ray is reporting a hit as if the `Platform` were at position `(0, 0)`

Please pardon me, I realize this may be working as expected, so maybe I'm asking for some sort of enhancement? I basically followed a 2D Unity tutorial using raycasts and ported that over to Godot using `IntersectRay()` and encountered this situation.

**Steps to reproduce:**
Please see minimal reproduction project.

**Minimal reproduction project:**
[raycast-sub-frame-check.zip](https://github.com/godotengine/godot/files/1822793/raycast-sub-frame-check.zip)

**Comment by SirRamEsq** [comment]
I'm having the same issue as this.  Making use of `force_raycast_update` doesn't seem to alleviate the problem either.
I use raycasts for moving across terrain in a 2D platformer. As long as the colliding body against the raycast is a `StaticBody`, my collisions are fine. With any moving scenes, however, the reported collision points lag one frame behind.

**Comment by SirRamEsq** [comment]
Digging around in the source code for a bit, I found a comment that reads
"//not quite correct, should compute the next matrix.."
In collision_object_2d_sw.cpp 
CollisionObject2DSW::_update_shapes()

Is this comment implying that updating collisions for all CollisionObject2D is a frame behind?

**Comment by SirRamEsq** [comment]
Not sure what the fix might be for this (Or even if it's actually a problem or if it's just us). I did come up with a work around for my use case.

I created a scene called "RayCastPosBugFix". It's just a `Node2D` with this as its script (Please excuse the camelCase styling...)
```
extends Node2D

# This class will measure how much its global_position has changed over the course of one frame
# Attach to parent and keep local transform the same in order to measure the parent's change in global_position
# Add the translationDifference to the raycast point in order to get the correct raycast position

var fTranslationDifference = 0.0
var _fPreviousFrameGlobal = 0.0

func _ready():
	_fPreviousFrameGlobal = global_position

func _physics_process(delta):
	fTranslationDifference = global_position - _fPreviousFrameGlobal
	_fPreviousFrameGlobal = global_position
```

I then instance this scene as a child to all moving physics bodies

Then, when performing a `RayCast`, I do the following:

```
# If body does not move bug fix doesn't need to be applied
if collider is StaticBody2D:
	set_global_position(newPos)
else:
	if collider.has_node("RayCastPosBugFix"):
		var bugFix = collider.find_node("RayCastPosBugFix")
		newPos += bugFix.fTranslationDifference
	set_global_position(newPos)
```

Hope this can help someone out.

**Comment by SirRamEsq** [comment]
After toying with this some more, I've noticed this bug only occurs 'sometimes'
With certain scenes, Raycast2Ds will always report the correct position of the colliding scene
With certain other scenes, they will always report the previous frame's position

One instance of a scene can always be correct, whereas another identical instance of the scene will always lag one frame behind.

It leads me to believe that perhaps this bug is caused by the order in which scenes are updated.
If a scene that would collide with a RayCast has already had its movement applied, then the RayCast will correctly intersect with it
Whereas if the scene in question will be moved after the intersect_ray is called, then the RayCast will lag one frame behind

**Comment by eon-s** [comment]
Have you tried moving the transform.origin of the body state instead? setting position may not update the body state as fast as you want (or at all in some cases).

**Comment by SirRamEsq** [comment]
Not entirely sure how to set the origin correctly vs setting the position

Currently, I update scenes by what is intuitive to me.
For the player, I use a KinematicBody2D and call `move_and_slide,` in addition to manually setting the position afterward via a few RayCast2Ds
When setting the position manually, I'll use either `position` or `global_position`

Some scenes (like kinematicBodies acting as floating platforms) do not set their own position or call `move_and_slide`, but are moved by parent nodes that rotate, follow a path, or move linearly

When landing on a floating platform that moves, the player RayCasting against the moving platform will report an incorrect position for certain combinations of player and moving platform (which I think is based on the update order).

How would setting the transform.origin help? What's the difference between setting the position vs setting the origin?

**Comment by eon-s** [comment]
Setting position does not affect the body state immediately and is possible that rays are failing for that reason.
If that is the case, then bodies may need a better method to sync the state to the Node transform

To modify the state's transform, use (the C# equivalent of) `Physics2DServer.body_get_direct_state` for the state, then just `state.transform.origin` to move it.

I suspect that this is what may be happening and the way to deal with this could have room for improvement.

**Comment by reduz** [comment]
This is not a bug, for performance reasons you should not expect objects to update their position in the physics immediately after you move them. Transforms are accumulated and then set.

Probably you could try call_deferred after you move an object if you want to do a raycast.

**Comment by reduz** [comment]
I was thinking, maybe a method to force an object update could be added, since this is not the first time it is asked for.

**Comment by reduz** [comment]
fixed via 74359a1d1
